### ISSUE TITLE
Race between tablet split and repair

### ISSUE THREAD

**Issue by raphaelsc** [scylladb/scylladb#19378]
Consider the following:

```
T
0   split prepare starts
1                               repair starts
2   split prepare finishes
3                               repair adds unsplit sstables
4                               repair ends
5   split executes
```

1) If repair starts *during* 'split prepare', prepare can wait on unsplit sstables being written. But it can theoretically happen *no* sstable is being written (by repair) on this shard now, but it will soon after 'split prepare finishes'.

2) If repair starts *after* 'split prepare' and *before* 'split execution', prepare already completed and will miss unsplit sstables.

Requirement: We want split of tablets to happen in parallel to migration / repair / load-and-stream / whatever.

What we could do is to serialize 'split prepare finish' with e.g. repair. But maybe we'd need repair to move to group0 state, such that replicas can verify that no global write process is ongoing, before marking themselves as 'split prepared'.

**Comment by raphaelsc** [comment]
@tgrabiec please comment.

**Comment by raphaelsc** [comment]
Alternatively, we do it all in replica. 'split prepare' waits for ongoing writes, and new ones will use the mutation splitter (used by split compaction).

**Comment by raphaelsc** [comment]
@tgrabiec I suggest we fix it like this

```
diff --git a/replica/table.cc b/replica/table.cc
index 1493d54c06..2a6c6659e2 100644
--- a/replica/table.cc
+++ b/replica/table.cc
@@ -1046,7 +1046,7 @@ table::do_add_sstable_and_update_cache(sstables::shared_sstable sst, sstables::o
     auto holder = cg.async_gate().hold();
 
     auto permit = co_await seastar::get_units(_sstable_set_mutation_sem, 1);
-    co_return co_await get_row_cache().invalidate(row_cache::external_updater([&] () noexcept {
+    co_await get_row_cache().invalidate(row_cache::external_updater([&] () noexcept {
         // FIXME: this is not really noexcept, but we need to provide strong exception guarantees.
         // atomically load all opened sstables into column family.
         if (!offstrategy) {
@@ -1059,6 +1059,8 @@ table::do_add_sstable_and_update_cache(sstables::shared_sstable sst, sstables::o
             try_trigger_compaction(cg);
         }
     }), dht::partition_range::make({sst->get_first_decorated_key(), true}, {sst->get_last_decorated_key(), true}));
+
+    co_await maybe_split_compaction_group_of(locator::tablet_id(cg.group_id()));
 }
 
```

that's essentially the same way when we guarantee split of a tablet that is being migrated during prepare phase.

by waiting on that split, we guarantee repair won't move on until new sstable produced is split, and given that repair hols erm, that will synchronize with split execution. what do you think?

**Comment by temichus** [comment]
last time seen in SCT tier1 2024-07-06

**Comment by enaydanov** [comment]
Another related crash (similar to https://github.com/scylladb/scylladb/issues/19872):

```
2024-07-24T08:15:40.506+00:00 longevity-50gb-12h-master-db-node-a7ad09ee-1     !INFO | scylla[10060]: Aborting on shard 0, in scheduling group main.
void seastar::backtrace<seastar::backtrace_buffer::append_backtrace()::{lambda(seastar::frame)#1}>(seastar::backtrace_buffer::append_backtrace()::{lambda(seastar::frame)#1}&&) at reactor.cc:?
seastar::print_with_backtrace(seastar::backtrace_buffer&, bool) at reactor.cc:?
seastar::install_oneshot_signal_handler<6, (void (*)())(&seastar::sigabrt_action)>()::{lambda(int, siginfo_t*, void*)#1}::__invoke(int, siginfo_t*, void*) at reactor.cc:?
?? ??:0
?? ??:0
?? ??:0
?? ??:0
seastar::on_internal_error(seastar::logger&, std::basic_string_view<char, std::char_traits<char> >) at on_internal_error.cc:?
replica::tablet_storage_group_manager::compaction_group_for_sstable(seastar::lw_shared_ptr<sstables::sstable> const&) const at table.cc:?
compaction::shard_reshaping_compaction_task_impl::run() at task_manager_module.cc:?
tasks::task_manager::task::impl::run_to_completion() at task_manager.cc:?
tasks::task_manager::task::start() at task_manager.cc:?
_ZN5tasks12task_manager6module19make_and_start_taskIN10compaction36shard_reshaping_compaction_task_implEJRNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESB_RN5utils11tagged_uuidINS_11task_id_tagEEERN8sstables17sstable_directoryERN7seastar7shardedIN7replica8databaseEEERNSH_12reshape_modeERSt8functionIFNSK_13lw_shared_ptrINSH_7sstableEEEjEERSS_IFbRKSV_EERmEQaasr3stdE12is_base_of_vINS0_4task4implET_ErQNSK_10shared_ptrIS1_EEDpOT0__XcvS17__fL0p_spclsr3stdE7forwardIS1A_EfL0p0_EERSt7same_asIS17_EEEENSK_6futureINST_IS15_EEEENS_9task_infoES1C_ at task_manager_module.cc:?
compaction::table_reshaping_compaction_task_impl::run()::$_0::operator()(sstables::sstable_directory&) const at task_manager_module.cc:?
compaction::table_reshaping_compaction_task_impl::run() at task_manager_module.cc:?
tasks::task_manager::task::impl::run_to_completion() at task_manager.cc:?
tasks::task_manager::task::start() at task_manager.cc:?
_ZN5tasks12task_manager6module19make_and_start_taskIN10compaction36table_reshaping_compaction_task_implEJN7seastar13basic_sstringIcjLj15ELb1EEES7_RNS5_7shardedIN8sstables17sstable_directoryEEERNS8_IN7replica8databaseEEERNS9_12reshape_modeESt8functionIFNS5_13lw_shared_ptrINS9_7sstableEEEjEESJ_IFbRKSM_EEEQaasr3stdE12is_base_of_vINS0_4task4implET_ErQNS5_10shared_ptrIS1_EEDpOT0__XcvSV__fL0p_spclsr3stdE7forwardISY_EfL0p0_EERSt7same_asISV_EEEENS5_6futureINSK_IST_EEEENS_9task_infoES10_ at distributed_loader.cc:?
replica::distributed_loader::reshape(seastar::sharded<sstables::sstable_directory>&, seastar::sharded<replica::database>&, sstables::reshape_mode, seastar::basic_sstring<char, unsigned int, 15u, true>, seastar::basic_sstring<char, unsigned int, 15u, true>, std::function<seastar::lw_shared_ptr<sstables::sstable> (unsigned int)>, std::function<bool (seastar::lw_shared_ptr<sstables::sstable> const&)>) at distributed_loader.cc:?
replica::table_populator::populate_subdir(sstables::sstable_state, seastar::bool_class<replica::allow_offstrategy_compaction_tag>) [clone .resume] at distributed_loader.cc:?
seastar::internal::coroutine_traits_base<void>::promise_type::run_and_dispose() at main.cc:?
seastar::reactor::run_tasks(seastar::reactor::task_queue&) at reactor.cc:?
seastar::reactor::run_some_tasks() at reactor.cc:?
seastar::reactor::do_run() at reactor.cc:?
seastar::reactor::run() at reactor.cc:?
seastar::app_template::run_deprecated(int, char**, std::function<void ()>&&) at app-template.cc:?
seastar::app_template::run(int, char**, std::function<seastar::future<int> ()>&&) at app-template.cc:?
scylla_main(int, char**) at main.cc:?
std::function<int (int, char**)>::operator()(int, char**) const at ??:?
main at main.cc:?
?? ??:0
?? ??:0
_start at ??:?
```

## Packages

Scylla version: `6.2.0~dev-20240723.bac7c3331377` with build-id `893ecd893b4dae1d14bda76631e7c9cfc34f0123`

Kernel Version: `6.8.0-1011-aws`

## Installation details

Cluster size: 4 nodes (i4i.4xlarge)

Scylla Nodes used in this run:
- longevity-50gb-12h-master-db-node-a7ad09ee-4 (34.241.14.67 | 10.4.8.238) (shards: 10)
- longevity-50gb-12h-master-db-node-a7ad09ee-3 (63.35.191.46 | 10.4.10.182) (shards: 12)
- longevity-50gb-12h-master-db-node-a7ad09ee-2 (34.242.229.241 | 10.4.9.11) (shards: 10)
- longevity-50gb-12h-master-db-node-a7ad09ee-1 (34.245.142.234 | 10.4.10.121) (shards: 12)


OS / Image: `ami-02a662ef3c73ef553` (aws: undefined_region)

Test: `longevity-150gb-asymmetric-cluster-12h-test`
Test id: `a7ad09ee-1223-4664-9f9d-1f9135c57be9`
Test name: `scylla-master/tier1/longevity-150gb-asymmetric-cluster-12h-test`
Test config file(s):

- [longevity-150GB-12h-autorization-LimitedMonkey.yaml](https://github.com/scylladb/scylla-cluster-tests/blob/e87674d93edb4c28da21a30428470a16ca8d462c/test-cases/longevity/longevity-150GB-12h-autorization-LimitedMonkey.yaml)


<details>
<summary>
Logs and commands
</summary>


- Restore Monitor Stack command: `$ hydra investigate show-monitor a7ad09ee-1223-4664-9f9d-1f9135c57be9`
- Restore monitor on AWS instance using [Jenkins job](https://jenkins.scylladb.com/view/QA/job/QA-tools/job/hydra-show-monitor/parambuild/?test_id=a7ad09ee-1223-4664-9f9d-1f9135c57be9)
- Show all stored logs command: `$ hydra investigate show-logs a7ad09ee-1223-4664-9f9d-1f9135c57be9`


## Logs:
- **core.scylla-longevity-50gb-12h-master-db-node-a7ad09ee-2-2024-07-24_08-04-38.gz** - [https://storage.cloud.google.com/upload.scylladb.com/core.scylla.106.51e253d7ff50488996cb1f293039caa4.8695.1721808006000000./core.scylla.106.51e253d7ff50488996cb1f293039caa4.8695.1721808006000000.zst](https://storage.cloud.google.com/upload.scylladb.com/core.scylla.106.51e253d7ff50488996cb1f293039caa4.8695.1721808006000000./core.scylla.106.51e253d7ff50488996cb1f293039caa4.8695.1721808006000000.zst)
- **core.scylla-longevity-50gb-12h-master-db-node-a7ad09ee-1-2024-07-24_08-05-04.gz** - [https://storage.cloud.google.com/upload.scylladb.com/core.scylla.106.ba94d9f130f9408eaa66ce4378b566cc.9230.1721808038000000./core.scylla.106.ba94d9f130f9408eaa66ce4378b566cc.9230.1721808038000000.zst](https://storage.cloud.google.com/upload.scylladb.com/core.scylla.106.ba94d9f130f9408eaa66ce4378b566cc.9230.1721808038000000./core.scylla.106.ba94d9f130f9408eaa66ce4378b566cc.9230.1721808038000000.zst)
- **core.scylla-longevity-50gb-12h-master-db-node-a7ad09ee-3-2024-07-24_08-05-24.gz** - [https://storage.cloud.google.com/upload.scylladb.com/core.scylla.106.6195fac67f89438a8cea87dc8d2421d0.8819.1721808036000000./core.scylla.106.6195fac67f89438a8cea87dc8d2421d0.8819.1721808036000000.zst](https://storage.cloud.google.com/upload.scylladb.com/core.scylla.106.6195fac67f89438a8cea87dc8d2421d0.8819.1721808036000000./core.scylla.106.6195fac67f89438a8cea87dc8d2421d0.8819.1721808036000000.zst)
- **core.scylla-longevity-50gb-12h-master-db-node-a7ad09ee-2-2024-07-24_08-08-19.gz** - [https://storage.cloud.google.com/upload.scylladb.com/core.scylla.106.51e253d7ff50488996cb1f293039caa4.8957.1721808233000000./core.scylla.106.51e253d7ff50488996cb1f293039caa4.8957.1721808233000000.zst](https://storage.cloud.google.com/upload.scylladb.com/core.scylla.106.51e253d7ff50488996cb1f293039caa4.8957.1721808233000000./core.scylla.106.51e253d7ff50488996cb1f293039caa4.8957.1721808233000000.zst)
- **core.scylla-longevity-50gb-12h-master-db-node-a7ad09ee-3-2024-07-24_08-08-39.gz** - [https://storage.cloud.google.com/upload.scylladb.com/core.scylla.106.6195fac67f89438a8cea87dc8d2421d0.9084.1721808257000000./core.scylla.106.6195fac67f89438a8cea87dc8d2421d0.9084.1721808257000000.zst](https://storage.cloud.google.com/upload.scylladb.com/core.scylla.106.6195fac67f89438a8cea87dc8d2421d0.9084.1721808257000000./core.scylla.106.6195fac67f89438a8cea87dc8d2421d0.9084.1721808257000000.zst)
- **core.scylla-longevity-50gb-12h-master-db-node-a7ad09ee-1-2024-07-24_08-08-49.gz** - [https://storage.cloud.google.com/upload.scylladb.com/core.scylla.106.ba94d9f130f9408eaa66ce4378b566cc.9504.1721808264000000./core.scylla.106.ba94d9f130f9408eaa66ce4378b566cc.9504.1721808264000000.zst](https://storage.cloud.google.com/upload.scylladb.com/core.scylla.106.ba94d9f130f9408eaa66ce4378b566cc.9504.1721808264000000./core.scylla.106.ba94d9f130f9408eaa66ce4378b566cc.9504.1721808264000000.zst)
- **core.scylla-longevity-50gb-12h-master-db-node-a7ad09ee-2-2024-07-24_08-11-29.gz** - [https://storage.cloud.google.com/upload.scylladb.com/core.scylla.106.51e253d7ff50488996cb1f293039caa4.9154.1721808460000000./core.scylla.106.51e253d7ff50488996cb1f293039caa4.9154.1721808460000000.zst](https://storage.cloud.google.com/upload.scylladb.com/core.scylla.106.51e253d7ff50488996cb1f293039caa4.9154.1721808460000000./core.scylla.106.51e253d7ff50488996cb1f293039caa4.9154.1721808460000000.zst)
- **core.scylla-longevity-50gb-12h-master-db-node-a7ad09ee-3-2024-07-24_08-12-23.gz** - [https://storage.cloud.google.com/upload.scylladb.com/core.scylla.106.6195fac67f89438a8cea87dc8d2421d0.9271.1721808478000000./core.scylla.106.6195fac67f89438a8cea87dc8d2421d0.9271.1721808478000000.zst](https://storage.cloud.google.com/upload.scylladb.com/core.scylla.106.6195fac67f89438a8cea87dc8d2421d0.9271.1721808478000000./core.scylla.106.6195fac67f89438a8cea87dc8d2421d0.9271.1721808478000000.zst)
- **core.scylla-longevity-50gb-12h-master-db-node-a7ad09ee-1-2024-07-24_08-12-37.gz** - [https://storage.cloud.google.com/upload.scylladb.com/core.scylla.106.ba94d9f130f9408eaa66ce4378b566cc.9697.1721808487000000./core.scylla.106.ba94d9f130f9408eaa66ce4378b566cc.9697.1721808487000000.zst](https://storage.cloud.google.com/upload.scylladb.com/core.scylla.106.ba94d9f130f9408eaa66ce4378b566cc.9697.1721808487000000./core.scylla.106.ba94d9f130f9408eaa66ce4378b566cc.9697.1721808487000000.zst)
- **core.scylla-longevity-50gb-12h-master-db-node-a7ad09ee-2-2024-07-24_08-15-10.gz** - [https://storage.cloud.google.com/upload.scylladb.com/core.scylla.106.51e253d7ff50488996cb1f293039caa4.9322.1721808632000000./core.scylla.106.51e253d7ff50488996cb1f293039caa4.9322.1721808632000000.zst](https://storage.cloud.google.com/upload.scylladb.com/core.scylla.106.51e253d7ff50488996cb1f293039caa4.9322.1721808632000000./core.scylla.106.51e253d7ff50488996cb1f293039caa4.9322.1721808632000000.zst)
- **core.scylla-longevity-50gb-12h-master-db-node-a7ad09ee-3-2024-07-24_08-16-08.gz** - [https://storage.cloud.google.com/upload.scylladb.com/core.scylla.106.6195fac67f89438a8cea87dc8d2421d0.9445.1721808698000000./core.scylla.106.6195fac67f89438a8cea87dc8d2421d0.9445.1721808698000000.zst](https://storage.cloud.google.com/upload.scylladb.com/core.scylla.106.6195fac67f89438a8cea87dc8d2421d0.9445.1721808698000000./core.scylla.106.6195fac67f89438a8cea87dc8d2421d0.9445.1721808698000000.zst)
- **core.scylla-longevity-50gb-12h-master-db-node-a7ad09ee-1-2024-07-24_08-16-25.gz** - [https://storage.cloud.google.com/upload.scylladb.com/core.scylla.106.ba94d9f130f9408eaa66ce4378b566cc.9873.1721808713000000./core.scylla.106.ba94d9f130f9408eaa66ce4378b566cc.9873.1721808713000000.zst](https://storage.cloud.google.com/upload.scylladb.com/core.scylla.106.ba94d9f130f9408eaa66ce4378b566cc.9873.1721808713000000./core.scylla.106.ba94d9f130f9408eaa66ce4378b566cc.9873.1721808713000000.zst)
- **core.scylla-longevity-50gb-12h-master-db-node-a7ad09ee-2-2024-07-24_08-18-49.gz** - [https://storage.cloud.google.com/upload.scylladb.com/core.scylla.106.51e253d7ff50488996cb1f293039caa4.9494.1721808862000000./core.scylla.106.51e253d7ff50488996cb1f293039caa4.9494.1721808862000000.zst](https://storage.cloud.google.com/upload.scylladb.com/core.scylla.106.51e253d7ff50488996cb1f293039caa4.9494.1721808862000000./core.scylla.106.51e253d7ff50488996cb1f293039caa4.9494.1721808862000000.zst)
- **core.scylla-longevity-50gb-12h-master-db-node-a7ad09ee-3-2024-07-24_08-19-35.gz** - [https://storage.cloud.google.com/upload.scylladb.com/core.scylla.106.6195fac67f89438a8cea87dc8d2421d0.9610.1721808919000000./core.scylla.106.6195fac67f89438a8cea87dc8d2421d0.9610.1721808919000000.zst](https://storage.cloud.google.com/upload.scylladb.com/core.scylla.106.6195fac67f89438a8cea87dc8d2421d0.9610.1721808919000000./core.scylla.106.6195fac67f89438a8cea87dc8d2421d0.9610.1721808919000000.zst)
- **db-cluster-a7ad09ee.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/a7ad09ee-1223-4664-9f9d-1f9135c57be9/20240724_082222/db-cluster-a7ad09ee.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/a7ad09ee-1223-4664-9f9d-1f9135c57be9/20240724_082222/db-cluster-a7ad09ee.tar.gz)
- **sct-runner-events-a7ad09ee.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/a7ad09ee-1223-4664-9f9d-1f9135c57be9/20240724_082222/sct-runner-events-a7ad09ee.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/a7ad09ee-1223-4664-9f9d-1f9135c57be9/20240724_082222/sct-runner-events-a7ad09ee.tar.gz)
- **sct-a7ad09ee.log.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/a7ad09ee-1223-4664-9f9d-1f9135c57be9/20240724_082222/sct-a7ad09ee.log.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/a7ad09ee-1223-4664-9f9d-1f9135c57be9/20240724_082222/sct-a7ad09ee.log.tar.gz)
- **loader-set-a7ad09ee.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/a7ad09ee-1223-4664-9f9d-1f9135c57be9/20240724_082222/loader-set-a7ad09ee.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/a7ad09ee-1223-4664-9f9d-1f9135c57be9/20240724_082222/loader-set-a7ad09ee.tar.gz)
- **monitor-set-a7ad09ee.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/a7ad09ee-1223-4664-9f9d-1f9135c57be9/20240724_082222/monitor-set-a7ad09ee.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/a7ad09ee-1223-4664-9f9d-1f9135c57be9/20240724_082222/monitor-set-a7ad09ee.tar.gz)


[Jenkins job URL](https://jenkins.scylladb.com/job/scylla-master/job/tier1/job/longevity-150gb-asymmetric-cluster-12h-test/26/)
[Argus](https://argus.scylladb.com/test/7f604213-5cad-4418-921e-1ca98918071c/runs?additionalRuns[]=a7ad09ee-1223-4664-9f9d-1f9135c57be9)
</details>

**Comment by tgrabiec** [comment]
> @bhalevy , @raphaelsc - if this is not a regression from 6.0, I suggest we defer it to 6.2 (and backport to 6.1 when ready)

It's not regression in tablets, but it's a regression compared to vnodes. If tablets are not experimental, it should be treated as regression IMO.

**Comment by mykaul** [comment]
> > @bhalevy , @raphaelsc - if this is not a regression from 6.0, I suggest we defer it to 6.2 (and backport to 6.1 when ready)
> 
> It's not regression in tablets, but it's a regression compared to vnodes. If tablets are not experimental, it should be treated as regression IMO.

The PR hasn't made much progress. I'm pushing this issue to 6.2, in the hope we can fix it quickly for 6.1.1.

**Comment by raphaelsc** [comment]
I resumed my work on it.
